singularity-eos/eos/get_sg_eos.cpp

```diff
@@ -0,0 +1,491 @@
+#else  // KOKKOS_ENABLE_SERIAL
+  // assume atomics are required for correctness if serial not even enabled
+  constexpr auto at_int{at_int_full};
+#endif // KOKKOS_ENABLE_SERIAL
+  // set up scratch arrays
+  constexpr auto KGlobal = Kokkos::Experimental::UniqueTokenScope::Global;
+  Kokkos::Experimental::UniqueToken<DES, KGlobal> tokens{};
+  using VAWI = Kokkos::ViewAllocateWithoutInitializing;
+
+  const bool small_loop{tokens.size() > ncell};
+  const decltype(tokens)::size_type scratch_size{std::min(tokens.size(), ncell)};
+  ScratchV<int> pte_mats(VAWI("PTE::scratch mats"), scratch_size, nmat);
+  ScratchV<int> pte_idxs(VAWI("PTE::scratch idxs"), scratch_size, nmat);
+  ScratchV<double> mass_pte(VAWI("PTE::scratch mass"), scratch_size, nmat);
+  ScratchV<double> sie_pte(VAWI("PTE::scratch sie"), scratch_size, nmat);
+  ScratchV<double> vfrac_pte(VAWI("PTE::scratch vfrac"), scratch_size, nmat);
+  ScratchV<double> temp_pte(VAWI("PTE::scratch temp"), scratch_size, nmat);
+  ScratchV<double> press_pte(VAWI("PTE::scratch press"), scratch_size, nmat);
+  ScratchV<double> rho_pte(VAWI("PTE::scratch rho"), scratch_size, nmat);
+  // declare solver scratch, allocate in the case statement
+  int pte_solver_scratch_size{};
+  ScratchV<double> solver_scratch;
+
+  // create helper lambdas to reduce code duplication
+  const auto init_lambda = SG_GET_SG_EOS_INIT_LAMBDA_DECL;
+  const auto final_lambda = SG_GET_SG_EOS_FINAL_LAMBDA_DECL;
+  Kokkos::View<int, MemoryTraits<at_int>> res("PTE::num fails");
+  Kokkos::View<int, MemoryTraits<at_int>> n_solves("PTE::num solves");
+  const std::string perf_nums =
+      "[" + std::to_string(nmat) + "," + std::to_string(ncell) + "]";
+  auto input_int_enum = static_cast<input_condition>(input_int);
+  switch (input_int_enum) {
+  case input_condition::RHO_T_INPUT: {
+    // T-rho input
+    // set frac_vol = 1/nmat
+    // set rho_i = nmat / spvol * frac_mass_i
+    // iterate PTE solver to obtain internal energies
+    // that results in the input T
+    pte_solver_scratch_size = PTESolverFixedTRequiredScratch(nmat);
+    solver_scratch = ScratchV<double>(VAWI("PTE::scratch solver"), scratch_size,
+                                      pte_solver_scratch_size);
+    const std::string rt_name = "PTE::solve (rho,T) input" + perf_nums;
+    singularity::get_sg_eos_rho_t(rt_name.c_str(), ncell, nmat, offsets_v,
+                                  eos_offsets_v, eos_v, press_v, pmax_v, vol_v,
+                                  spvol_v, sie_v, temp_v, bmod_v, dpde_v, cv_v,
+                                  frac_mass_v, frac_vol_v, frac_ie_v,
+                                  frac_bmod_v, frac_dpde_v, frac_cv_v, 
+                                  pte_idxs, pte_mats, press_pte, vfrac_pte,
+                                  rho_pte, sie_pte, temp_pte, solver_scratch,
+                                  tokens, small_loop, do_frac_bmod,
+                                  do_frac_dpde, do_frac_cv);
+//    portableFor(
```

Comment:
maybe remove the comments?

Reply:
still draft, just moving around for more testing.